--- mm/fremap.c.original	2014-10-20 22:21:10.924201325 +0000
+++ mm/fremap.c	2014-10-20 22:23:41.827205429 +0000
@@ -202,15 +202,26 @@
 		 */
 		if (mapping_cap_account_dirty(mapping)) {
 			unsigned long addr;
-			struct file *file = get_file(vma->vm_file);
+			struct file *file = vma->vm_file,
+                               *prfile = vma->vm_prfile;
 
+			vma_get_file(vma);
 			addr = mmap_region(file, start, size,
 					vma->vm_flags, pgoff);
-			fput(file);
+			vma_fput(vma);
 			if (IS_ERR_VALUE(addr)) {
 				err = addr;
 			} else {
 				BUG_ON(addr != start);
+				if (prfile) {
+                                       struct vm_area_struct *new_vma;
+
+                                       new_vma = find_vma(mm, addr);
+                                       if (!new_vma->vm_prfile)
+                                               new_vma->vm_prfile = prfile;
+                                       if (new_vma != vma)
+                                               get_file(prfile);
+                                }
 				err = 0;
 			}
 			goto out;
